src/api/data_source_routes.ts

```diff
@@ -295,5 +303,60 @@ export default class DataSourceRoutes {
             .catch((err) => res.status(500).send(err))
             .finally(() => next())
     }
+
+    private static uploadFile(req: Request, res: Response, next: NextFunction) {
+        const fileNames: string[] = []
+        const busboy = new Busboy({headers: req.headers})
+        const metadata: {[key: string]: any} = {}
+        let metadataFieldCount = 0;
+
+        // upload the file to the relevant file storage provider, saving the file name
+        // we can't actually wait on the full upload to finish, so there is no way we
+        // can take information about the upload and pass it later on in the busboy parsing
+        // because of this we're treating the file upload as fairly standalone
+        busboy.on('file', (fieldname: string, file: NodeJS.ReadableStream, filename: string, encoding: string, mimeType: string) => {
+            const user = req.user as UserT
+            DataSourceUploadFile(req.params.id, req.params.sourceID, user.id!, filename, encoding, mimeType, file as Readable)
```

Comment:
`file as Readable` must be there as the type system doesn't recognize that `Readable` implements the `NodeJS.ReadableStream` interface correctly.